### PR TITLE
Lint inheritence

### PR DESCRIPTION
**Pull request by haim-kermany**
some reorg code, to remove redundant function
still need to delete some files, that became too small.
will do it in different PR, to make this one much easier to review.

### CODE REVIEW

**Review by zivnevo — CHANGES_REQUESTED** [review]
This looks great: it much improves readability and reduces LoC.
Just two small comments:
1. Let's be consistent about whether we use consts for linter name+description or not.
2. All constructors should use field specifiers when initializing the struct.

**Comment by haim-kermany** [comment]
> 2\. specifiers

fixed

**Comment by haim-kermany** [comment]
> This looks great: it much improves readability and reduces LoC. Just two small comments:
> 
>     1. Let's be consistent about whether we use consts for linter name+description or not.
> 
>     2. All constructors should use field specifiers when initializing the struct.

fixed

**Review by zivnevo — APPROVED** [review]

File: pkg/linter/linterTypes.go [hunk 1]
```diff
@@ -66,3 +79,18 @@ func (lint *basicLinter) toJSON() []any {
 	}
 	return res
 }
+
+type filterLinter struct {
+	basicLinter
+	layer          string
+	filterFindings func(map[string]*vpcmodel.VPCConfig, string) ([]finding, error)
```
Comment (ShiriMoran) [inline pkg/linter/linterTypes.go]: Cool, just `filterFinding...` is not the right name here. Renamed and pushed.

File: pkg/linter/lintRuleNonRelevantCIDRCommon.go [hunk 2]
```diff
@@ -25,7 +25,7 @@ type ruleNonRelevantCIDR struct {
 // functionality used by both SG and NACL lints
 ////////////////////////////////////////////////////////////////////////////////////////////
 
-func findRuleNonRelevantCIDR(configs map[string]*vpcmodel.VPCConfig, filterLayerName string) (res []ruleNonRelevantCIDR, err error) {
+func findRuleNonRelevantCIDR(configs map[string]*vpcmodel.VPCConfig, filterLayerName string) (res []finding, err error) {
```
Comment (ShiriMoran) [inline pkg/linter/lintRuleNonRelevantCIDRCommon.go:28]: Why this change?

**Review by ShiriMoran — COMMENTED** [review]
Looks great, just one comment.
I do need the constant names of the linters for the enable disable mechanism

File: pkg/linter/lintRuleNonRelevantCIDRCommon.go (hunk 2 above)
Comment (haim-kermany) [reply pkg/linter/lintRuleNonRelevantCIDRCommon.go:28]: all the checkForFilter function must have the same signeture

File: pkg/common/genericSet.go [hunk 3]
```diff
@@ -90,3 +90,12 @@ func MapValues[K comparable, V any](m map[K]V) []V {
 	}
 	return vals
 }
+func MapKeys[K comparable, V any](m map[K]V) []K {
```
Comment (zivnevo) [inline pkg/common/genericSet.go:93]: `GenericSet.AsList()` should use this function

File: pkg/linter/lintFilterRuleSplitSubnetNACL.go [hunk 4]
```diff
@@ -8,31 +8,16 @@ package linter
...
-	rulesSplitSubnetsFound, err := findSplitRulesSubnet(lint.configs, vpcmodel.NaclLayer)
-	if err != nil {
-		return err
-	}
-	for i := range rulesSplitSubnetsFound {
-		lint.addFinding(&rulesSplitSubnetsFound[i])
-	}
-	return nil
+func newFilterRuleSplitSubnetLintNACL(name string, configs map[string]*vpcmodel.VPCConfig,
+	nodesConn map[string]*vpcmodel.VPCConnectivity) linter {
```
Comment (zivnevo) [inline pkg/linter/lintFilterRuleSplitSubnetNACL.go:13]:
```suggestion
	_ map[string]*vpcmodel.VPCConnectivity) linter {
```
Emphasizes this linter doesn't need connectivity information

File: pkg/linter/linterExecute.go [hunk 5]
```diff
@@ -8,34 +8,53 @@ package linter
...
-			ruleNonRelevantCIDRSGName: true}
-	for _, disable := range disableList {
-		enableLinters[disable] = false
-	}
-	for _, enable := range enableList {
-		enableLinters[enable] = true
-	}
 	nodesConn := map[string]*vpcmodel.VPCConnectivity{}
 	for uid, vpcConfig := range configs {
 		nodesConnThisCfg, err := vpcConfig.GetVPCNetworkConnectivity(false, true)
```
Comment (zivnevo) [inline pkg/linter/linterExecute.go:60]: Better move to a separate function, `computeConnectivity()` or otherwise move to `generateLinters()`

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/linter/lintFilterRuleSplitSubnetNACL.go (hunk 4 above)
Comment (ShiriMoran) [reply pkg/linter/lintFilterRuleSplitSubnetNACL.go:13]: I will handle this comment

File: pkg/common/genericSet.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/common/genericSet.go:88]: added comment, and issue https://github.com/np-guard/vpc-network-config-analyzer/issues/763

Comment (haim-kermany) [reply pkg/common/genericSet.go:88]: fixed

File: pkg/linter/lintFilterRuleSplitSubnetNACL.go (hunk 4 above)
Comment (haim-kermany) [reply pkg/linter/lintFilterRuleSplitSubnetNACL.go]: fixed

File: pkg/linter/linterExecute.go (hunk 5 above)
Comment (haim-kermany) [reply pkg/linter/linterExecute.go:57]: fixed

**Review by zivnevo — APPROVED** [review]

**Review by ShiriMoran — APPROVED** [review]
